base/statistics.jl

```diff
@@ -0,0 +1,438 @@
+# This file is a part of Julia. License is MIT: https://julialang.org/license
+
+##### mean #####
+
+"""
+    mean(itr)
+
+Compute the mean of all elements in a collection.
+
+!!! note
+    If `itr` contains `NaN` or [`missing`](@ref) values, the result is also
+    `NaN` or `missing` (`missing` takes precedence if array contains both).
+    Use the [`skipmissing`](@ref) function to omit `missing` entries and compute the
+    mean of non-missing values.
+
+!!! compat "Julia 1.9"
+    This method requires at least Julia 1.9.
+    Support for older releases is provided by the Statistics package.
+
+# Examples
+```jldoctest
+julia> mean(1:20)
+10.5
+
+julia> mean([1, missing, 3])
+missing
+
+julia> mean(skipmissing([1, missing, 3]))
+2.0
+```
+"""
+mean(itr) = mean(identity, itr)
+
+"""
+    mean(f, itr)
+
+Apply the function `f` to each element of collection `itr` and take the mean.
+
+!!! compat "Julia 1.9"
+    This method requires at least Julia 1.9.
+    Support for older releases is provided by the Statistics package.
+
+# Examples
+```jldoctest
+julia> mean(√, [1, 2, 3])
+1.3820881233139908
+
+julia> mean([√1, √2, √3])
+1.3820881233139908
+```
+"""
+function mean(f, itr)
+    y = iterate(itr)
+    if y === nothing
+        return Base.mapreduce_empty_iter(f, +, itr,
+                                         Base.IteratorEltype(itr)) / 0
+    end
+    count = 1
+    value, state = y
+    f_value = f(value)/1
+    total = Base.reduce_first(+, f_value)
+    y = iterate(itr, state)
+    while y !== nothing
+        value, state = y
+        total += _mean_promote(total, f(value))
+        count += 1
+        y = iterate(itr, state)
+    end
+    return total/count
+end
+
+"""
+    mean(f, A::AbstractArray; dims)
+
+Apply the function `f` to each element of array `A` and take the mean over dimensions `dims`.
+
+!!! compat "Julia 1.9"
+    This method requires at least Julia 1.9.
+    Support for older releases is provided by the Statistics package.
+
+```jldoctest
+julia> mean(√, [1, 2, 3])
+1.3820881233139908
+
+julia> mean([√1, √2, √3])
+1.3820881233139908
+
+julia> mean(√, [1 2 3; 4 5 6], dims=2)
+2×1 Matrix{Float64}:
+ 1.3820881233139908
+ 2.2285192400943226
+```
+"""
+mean(f, A::AbstractArray; dims=:) = _mean(f, A, dims)
```

Comment:
If you want to be able to dispatch on a keyword, how strange would it be to build in a hook in for licensed piracy... something like:
```suggestion
mean(f, A::AbstractArray; dims=:, weights=nothing) = _mean(f, A, dims, weights)
```

Reply:
Yeah that's the kind of pattern I had in mind. But let's leave that question for later as moving Statistics out would already be a big achievement.